src/shogun/mathematics/linalg/Vector.cpp

```diff
@@ -0,0 +1,227 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/mathematics/linalg/Vector.h>
+#include <shogun/mathematics/linalg/GPUVectorImpl.h>
+
+#ifdef HAVE_CXX11
+
+namespace shogun
+{
+
+template<class T>
+Vector<T>::Vector()
+{
+	init();
+}
+
+template<class T>
+Vector<T>::Vector(SGVector<T> const &vector)
+{
+	init();
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector + vector.vlen, m_data.get());
+	m_len = vector.vlen;
+}
+
+template<class T>
+Vector<T>::Vector(Vector<T> const &vector)
+{
+	init();
+
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+									[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+}
+
+template<class T>
+Vector<T>::~Vector()
+{
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(SGVector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data.get());
```

Comment:
There is a problem here:
We definitely dont want the data to be copied here. We want the same data to be re-used. Otherwise using linalg operations will always cause a memory copy.
Therefore, if copying from a SGVector, we need to increase the reference counting in order to avoid problems. This means that `Vector` here and the base class `SGReferencesData` need to share the ref counter business.
NOW, that is a problem if we are doing all this in a separate class (no idea how to resolve atm) 
A further problem is the alignment of the data, because we need to copy data in order to re-use memory from SGVector
@vigsterkr @lambday discussion needed